modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionStorage.scala

```diff
@@ -1,92 +1,124 @@
-import cats.syntax.foldable._
-import cats.syntax.functor._
-import cats.syntax.option._
-import cats.syntax.order._
-import cats.syntax.set._
-import cats.syntax.show._
-import cats.syntax.traverse._
+import cats.syntax.all._
+import cats.{Order, Show}
 
 import scala.annotation.tailrec
-import scala.collection.immutable.SortedSet
+import scala.collection.immutable.{SortedMap, SortedSet}
 import scala.util.control.NoStackTrace
 
 import org.tessellation.dag.l1.domain.transaction.TransactionStorage._
 import org.tessellation.ext.collection.MapRefUtils._
 import org.tessellation.node.shared.domain.transaction.filter.Consecutive
+import org.tessellation.schema.SnapshotOrdinal
 import org.tessellation.schema.address.Address
 import org.tessellation.schema.transaction._
 import org.tessellation.security.Hashed
 import org.tessellation.security.hash.Hash
+import org.tessellation.syntax.sortedCollection._
 
+import derevo.cats.eqv
+import derevo.derive
+import eu.timepit.refined.auto._
 import eu.timepit.refined.types.numeric.NonNegLong
 import io.chrisdavenport.mapref.MapRef
 import org.typelevel.log4cats.slf4j.Slf4jLogger
 
 class TransactionStorage[F[_]: Async](
-  lastAccepted: MapRef[F, Address, Option[LastTransactionReferenceState]],
-  waitingTransactions: MapRef[F, Address, Option[NonEmptySet[Hashed[Transaction]]]],
+  transactionsR: MapRef[F, Address, Option[SortedMap[TransactionOrdinal, StoredTransaction]]],
   initialTransactionReference: TransactionReference
 ) {
 
   private val logger = Slf4jLogger.getLogger[F]
   private val transactionLogger = Slf4jLogger.getLoggerFromName[F](transactionLoggerName)
 
-  def getInitialTxRef: TransactionReference = initialTransactionReference
+  def getInitialTx: MajorityTx = MajorityTx(initialTransactionReference, SnapshotOrdinal.MinValue)
 
-  def getState(): F[(Map[Address, LastTransactionReferenceState], Map[Address, NonEmptySet[Hashed[Transaction]]])] =
-    (lastAccepted.toMap, waitingTransactions.toMap).mapN((_, _))
+  def getState: F[Map[Address, SortedMap[TransactionOrdinal, StoredTransaction]]] = transactionsR.toMap
 
-  def getLastAcceptedReference(source: Address): F[TransactionReference] =
-    lastAccepted(source).get.map {
-      _.map(_.ref).getOrElse(initialTransactionReference)
+  def getTransactions(source: Address): F[Option[SortedMap[TransactionOrdinal, StoredTransaction]]] = transactionsR(source).get
+
+  def getLastProcessedTransaction(source: Address): F[StoredTransaction] =
+    transactionsR(source).get.map {
+      _.flatMap(getLastProcessedTransaction).getOrElse(getInitialTx)
     }
 
-  def setLastAccepted(lastTxRefs: Map[Address, TransactionReference]): F[Unit] =
-    lastAccepted.clear >>
-      lastTxRefs.toList.traverse {
-        case (address, reference) => lastAccepted(address).set(Majority(reference).some)
-      }.void
+  def getLastProcessedTransaction(maybeStored: Option[SortedMap[TransactionOrdinal, StoredTransaction]]): StoredTransaction =
```

Comment:
Not sure if this and the following function should be a part of the internal storage implementation. Especially that you fetch the transactions and then take them back to this function inside of the storage just to extract the correct one. I would rather see it as a pure function in the companion object. Or just extract the last tx based on the `StoredTransaction` type wherever it's needed - you are exposing the `StoredTransaction` either way.